<a href="https://colab.research.google.com/github/msudhanshu10/Complete-Data-Science/blob/master/Feature_Engineering_Hot_Encoding.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [9]:
! ls
! cd drive


drive  sample_data


In [13]:
from google.colab import files
uploaded = files.upload()

Saving mercedesbenz.csv to mercedesbenz (1).csv


In [44]:
df= pd.read_csv('mercedesbenz (1).csv')


In [45]:
df. head()

ID       y  X0 X1  X2 X3 X4  ... X378 X379 X380  X382  X383  X384  X385
0   0  130.81   k  v  at  a  d  ...    0    0    0     0     0     0     0
1   6   88.53   k  t  av  e  d  ...    0    0    0     0     0     0     0
2   7   76.26  az  w   n  c  d  ...    0    0    0     1     0     0     0
3   9   80.62  az  t   n  f  d  ...    0    0    0     0     0     0     0
4  13   78.02  az  v   n  f  d  ...    0    0    0     0     0     0     0

[5 rows x 378 columns]

In [46]:
cols=['y', 'X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8']
df= df[cols]

In [47]:
df.head()

y  X0 X1  X2 X3 X4 X5 X6 X8
0  130.81   k  v  at  a  d  u  j  o
1   88.53   k  t  av  e  d  y  l  o
2   76.26  az  w   n  c  d  x  j  x
3   80.62  az  t   n  f  d  x  l  e
4   78.02  az  v   n  f  d  h  d  n

In [48]:
for col in df.columns:
  print(col,':', df[col].nunique(), 'labels')

y : 2545 labels
X0 : 47 labels
X1 : 27 labels
X2 : 44 labels
X3 : 7 labels
X4 : 4 labels
X5 : 29 labels
X6 : 12 labels
X8 : 25 labels


In [49]:
pd.get_dummies(df, drop_first=True).shape
# .get_dummies is used for hot encoding

(4209, 188)

188 columns will be created in this data set if we ll use Hot Encoding.
When number of features increase, it leads to curse of Dimensionallity.
So, it may impact our accuracy level.


So we will use one Hot Encoding for n most occuring labels

In [50]:
df.X1.value_counts().sort_values(ascending=False)

aa    833
s     598
b     592
l     590
v     408
r     251
i     203
a     143
c     121
o      82
w      52
z      46
u      37
e      33
m      32
t      31
h      29
f      23
y      23
j      22
n      19
k      17
p       9
g       6
d       3
q       3
ab      3
Name: X1, dtype: int64

In [51]:
top_10_X1= [x for x in df.X1.value_counts().sort_values(ascending=False).head(10).index]
top_10_X1

['aa', 's', 'b', 'l', 'v', 'r', 'i', 'a', 'c', 'o']

In [52]:
for label in top_10_X1:
  df[label]= np.where(df['X1']==label,1,0)


In [53]:
df

y  X0 X1  X2 X3 X4  X5 X6 X8  aa  s  b  l  v  r  i  a  c  o
0     130.81   k  v  at  a  d   u  j  o   0  0  0  0  1  0  0  0  0  0
1      88.53   k  t  av  e  d   y  l  o   0  0  0  0  0  0  0  0  0  0
2      76.26  az  w   n  c  d   x  j  x   0  0  0  0  0  0  0  0  0  0
3      80.62  az  t   n  f  d   x  l  e   0  0  0  0  0  0  0  0  0  0
4      78.02  az  v   n  f  d   h  d  n   0  0  0  0  1  0  0  0  0  0
...      ...  .. ..  .. .. ..  .. .. ..  .. .. .. .. .. .. .. .. .. ..
4204  107.39  ak  s  as  c  d  aa  d  q   0  1  0  0  0  0  0  0  0  0
4205  108.77   j  o   t  d  d  aa  h  h   0  0  0  0  0  0  0  0  0  1
4206  109.22  ak  v   r  a  d  aa  g  e   0  0  0  0  1  0  0  0  0  0
4207   87.48  al  r   e  f  d  aa  l  u   0  0  0  0  0  1  0  0  0  0
4208  110.85   z  r  ae  c  d  aa  g  w   0  0  0  0  0  1  0  0  0  0

[4209 rows x 19 columns]

In [54]:
cols=['X0', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8']
top_10=[]
for col in df[cols]:
  top_10.append([x for x in df[col].value_counts().sort_values(ascending=False).head(10).index])

In [55]:
top_10

[['z', 'ak', 'y', 'ay', 't', 'x', 'o', 'f', 'n', 'w'],
 ['as', 'ae', 'ai', 'm', 'ak', 'r', 'n', 's', 'f', 'e'],
 ['c', 'f', 'a', 'd', 'g', 'e', 'b'],
 ['d', 'a', 'b', 'c'],
 ['w', 'v', 'q', 'r', 'd', 's', 'n', 'p', 'm', 'i'],
 ['g', 'j', 'd', 'i', 'l', 'a', 'h', 'k', 'c', 'b'],
 ['j', 's', 'f', 'n', 'i', 'e', 'r', 'a', 'w', 'v']]

In [56]:
type(top_10[0])


list

In [57]:
i=0
for col in cols:
  for label in top_10[i]:
    df[col+'_'+label]= np.where(df[col]==label,1,0)
  i=i+1

In [58]:
df.shape

(4209, 80)

In [59]:
df.head()

y  X0 X1  X2 X3 X4 X5  ... X8_n X8_i  X8_e  X8_r  X8_a  X8_w  X8_v
0  130.81   k  v  at  a  d  u  ...    0    0     0     0     0     0     0
1   88.53   k  t  av  e  d  y  ...    0    0     0     0     0     0     0
2   76.26  az  w   n  c  d  x  ...    0    0     0     0     0     0     0
3   80.62  az  t   n  f  d  x  ...    0    0     1     0     0     0     0
4   78.02  az  v   n  f  d  h  ...    1    0     0     0     0     0     0

[5 rows x 80 columns]